feat(parser): add support for constraint comprehension

tohrnii opened the pull request:
Partly addressing #187 and #220

grjte commented on parser/src/parser/tests/mod.rs:
```diff
@@ -12,6 +12,7 @@ mod arithmetic_ops;
 mod boundary_constraints;
 mod comments;
 mod constants;
+mod constraint_comprehension;
```
I would probably prefer to put the tests in with the existing constraints tests, especially since it's handled differently for boundary vs. integrity constraints. We could use a section separator marking them as being related to comprehension

grjte commented on parser/src/ast/integrity_constraints.rs:
```diff
@@ -1,11 +1,16 @@
 #[derive(Debug, Eq, PartialEq)]
 pub enum IntegrityStmt {
     Constraint(ConstraintType, Option<Expression>),
+    ConstraintComprehension(
+        ConstraintType,
+        Option<Expression>,
+        ConstraintComprehensionContext,
+    ),
```
I believe the only difference between this and the one above is the `ConstraintcomprehensionContext`, so I would probably prefer to take the same approach we took for selectors and do the following:

```
pub enum IntegrityStmt {
    Constraint(ConstraintType, Option<ConstraintComprehensionContext>, Option<Expression>),
    Variable(Variable),
}
```

I would also prefer to put the constraint comprehension context before the selector expression, since I think this is the way the language syntax should be ordered, as I noted elsewhere

grjte reviewed (changes requested):
Looks good overall, thank you! Just a few comments about structure

tohrnii commented:
Thank you @grjte, I think maybe we should merge this PR without addressing some of your comments right now. Things left to do:
1. Combining `Constraint` and `ConstraintComprehension` variants for both integrity and boundary constraints.
2. Modifying the syntax for constraint comprehension with selectors.
The reason is that these will require quite a bit refactor of the code and will be a waste of time rebasing all the downstream branches. I'll open an issue to address your points in another PR if that's ok with you and come back to it soon after the IR work is done.

tohrnii commented on parser/src/ast/integrity_constraints.rs:
```diff
@@ -38,3 +43,5 @@ impl IntegrityConstraint {
     }
 }
+
+pub type ConstraintComprehensionContext = Vec<(Identifier, Iterable)>;
```
I've defined a combined `ComprehensionContext` type in `air-script-core`.

grjte reviewed (approved):
@tohrnii sure, that makes sense. It's ok to revisit after the other refactor PRs are merged. I left 2 final nits inline that it would be good to take care of before merging this one.

tohrnii commented on parser/src/ast/boundary_constraints.rs:
```diff
@@ -56,3 +57,45 @@ impl Display for Boundary {
+    access: NamedTraceAccess,
+    boundary: Boundary,
+    expr: Expression,
+    context: Vec<(Identifier, Iterable)>,
```
Thanks, fixed.
